File: proxy_checker/FileCache.py
import json
import os
import time
from typing import Any, Optional


class FileCache:
    """
    A simple file-based cache system with expiration.

    Attributes:
        file_path (str): Path to the cache file.

    Methods:
        write_cache(value: Any, expires_in: int) -> None:
            Writes a value to the cache with an expiration time.

        read_cache() -> Optional[Any]:
            Reads the value from the cache if it has not expired.
    """

    def __init__(self, file_path: str) -> None:
        """
        Initializes the FileCache with a file path.

        Args:
            file_path (str): The path to the cache file.
        """
        self.file_path = file_path

    def write_cache(self, value: Any, expires_in: int) -> None:
        """
        Writes a value to the cache with an expiration time.

        Args:
            value (Any): The value to be cached.
            expires_in (int): The expiration time in seconds.
        """
        # Ensure the parent directory exists
        os.makedirs(os.path.dirname(self.file_path), exist_ok=True)

        cache_data = {
            "value": value,
            "timestamp": time.time(),
            "expires_in": expires_in,
        }
        with open(self.file_path, "w") as cache_file:
            json.dump(cache_data, cache_file)

    def read_cache(self) -> Optional[Any]:
        """
        Reads the value from the cache if it has not expired.

        Returns:
            Optional[Any]: The cached value if valid, otherwise None.
        """
        if not os.path.exists(self.file_path):
            return None

        try:
            with open(self.file_path, "r") as cache_file:
                cache_data = json.load(cache_file)

            current_time = time.time()
            cache_time = cache_data["timestamp"]
            expires_in = cache_data["expires_in"]

            if current_time - cache_time > expires_in:
                return None

            return cache_data["value"]
        except Exception:
            return None

File: proxy_checker/FileCache.php
<?php

declare(strict_types=1);

namespace ProxyChecker;

/**
 * Simple file-based cache helper with expiration.
 */
class FileCache
{
  /**
   * Path to the cache file.
   *
   * @var string
   */
  private $filePath;

  public function __construct(string $filePath)
  {
    $this->filePath = $filePath;
  }

  /**
   * Write a value to the cache with expiration (seconds).
   *
   * @param mixed $value
   * @param int $expiresIn Seconds until expiration
   *
   * @throws \RuntimeException on write/encode failures
   */
  public function writeCache($value, int $expiresIn): void {
    $dir = dirname($this->filePath);
    if (!is_dir($dir)) {
      if (!mkdir($dir, 0777, true) && !is_dir($dir)) {
        throw new \RuntimeException('Unable to create cache directory: ' . $dir);
      }
    }

    $cacheData = [
      'value'      => $value,
      'timestamp'  => microtime(true),
      'expires_in' => $expiresIn,
    ];

    $json = json_encode($cacheData, JSON_UNESCAPED_SLASHES);
    if ($json === false) {
      throw new \RuntimeException('Failed to encode cache data to JSON: ' . json_last_error_msg());
    }

    $fp = fopen($this->filePath, 'c');
    if ($fp === false) {
      throw new \RuntimeException('Unable to open cache file for writing: ' . $this->filePath);
    }

    try {
      if (!flock($fp, LOCK_EX)) {
        throw new \RuntimeException('Unable to obtain exclusive lock for cache file.');
      }

      ftruncate($fp, 0);
      rewind($fp);
      fwrite($fp, $json);
      fflush($fp);
      flock($fp, LOCK_UN);
    } finally {
      fclose($fp);
    }
  }

  /**
   * Read cached value if not expired, otherwise return null.
   *
   * @return mixed|null
   */
  public function readCache() {
    if (!file_exists($this->filePath) || !is_readable($this->filePath)) {
      return null;
    }

    $fp = fopen($this->filePath, 'r');
    if ($fp === false) {
      return null;
    }

    try {
      if (!flock($fp, LOCK_SH)) {
        return null;
      }

      $contents = stream_get_contents($fp);
      flock($fp, LOCK_UN);
    } finally {
      fclose($fp);
    }

    if ($contents === false || $contents === '') {
      return null;
    }

    $data = json_decode($contents, true);
    if (!is_array($data) || !isset($data['timestamp'], $data['expires_in'], $data['value'])) {
      return null;
    }

    $current = microtime(true);
    if (($current - (float) $data['timestamp']) > (int) $data['expires_in']) {
      return null;
    }

    return $data['value'];
  }
}
